File: test-fixtures/languages/javascript/micro/auth_service.js
const { execSync } = require("child_process");
const mysql = require("mysql");

const db = mysql.createConnection({ host: "localhost", database: "auth" });

function verifyToken(token) {
  // Inline the SQL string at the call site so the rule's
  // arg-shape SQL-keyword constraint can fire syntactically.
  return db.query("SELECT user_id FROM tokens WHERE token = '" + token + "'");  // sink: SQL injection
}

function runAdminCommand(userId, cmd) {
  if (userId) {
    execSync("notify-admin " + cmd);  // sink: command injection
  }
}

module.exports = { verifyToken, runAdminCommand };

File: test-fixtures/languages/cpp/micro/auth_service.cpp
#include "auth_service.h"
#include <cstdlib>
#include <sqlite3.h>

namespace micro {

std::string verify_token(const std::string& token) {
    sqlite3* db;
    sqlite3_stmt* stmt;
    sqlite3_open("auth.db", &db);
    std::string query = "SELECT user_id FROM tokens WHERE token = '" + token + "'";
    sqlite3_prepare_v2(db, query.c_str(), -1, &stmt, nullptr); // sink: SQL injection
    std::string user_id;
    if (sqlite3_step(stmt) == SQLITE_ROW) {
        user_id = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 0));
    }
    sqlite3_close(db);
    return user_id;
}

void run_admin_command(const std::string& user_id, const std::string& cmd) {
    if (!user_id.empty()) {
        std::string full_cmd = "notify-admin " + cmd;
        system(full_cmd.c_str()); // sink: command injection
    }
}

}
